File: Python3/1721. Swapping Nodes in a Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        cur = head
        
        # get the previous node of the k-th node
        for i in range(1, k):
            cur = cur.next
        #print(cur.val)    
        first = cur
        pre = head
        
        # get the previous node of the backwards k-th node
        while cur.next:
            pre = pre.next
            cur = cur.next
        
        first.val, pre.val = pre.val, first.val
        return head

File: Java/1721. Swapping Nodes in a Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapNodes(ListNode head, int k) {
        int count = 0;
        ListNode cur = head;
        ListNode front = new ListNode();
        ListNode end = new ListNode();
        
        while (cur != null) {
            count++;
            if (end != null) {
                end = end.next;
            }
            
            if (count == k) {
                front = cur;
                end = head;
            }
            cur = cur.next;
        }
        
        int temp = front.val;
        front.val = end.val;
        end.val = temp;
        
        return head;
    }
}